libtenzir/src/table_slice.cpp

```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
+
+auto bitmap_or(std::shared_ptr<arrow::Buffer> x,
+               std::shared_ptr<arrow::Buffer> y)
+  -> std::shared_ptr<arrow::Buffer> {
+  if (not x || not y) {
+    return nullptr;
+  }
+  auto size = std::min(x->size(), y->size());
+  auto z = check(arrow::AllocateBuffer(size));
+  // TODO: Is this safe?
+  auto x_ptr = x->data();
+  auto y_ptr = y->data();
+  auto z_ptr = z->mutable_data();
+  for (auto i = int64_t{0}; i < size; ++i) {
+    z_ptr[i] = x_ptr[i] | y_ptr[i]; // NOLINT
+  }
+  return z;
+}
+
+void unflatten_recurse(unflatten_entry& root, const arrow::StructArray& array,
+                       std::string_view sep) {
+  auto ty = array.struct_type();
+  auto names = ty->fields() | std::views::transform(&arrow::Field::name);
+  // We need to flatten the null bitmap here because it can happen that the
+  // fields are saved to a record that is made non-null by another entry.
+  auto fields = check(array.Flatten());
+  if (not std::holds_alternative<unflatten_record>(root)) {
+    root.emplace<unflatten_record>(array.length(), array.null_bitmap());
+  }
+  for (auto [name, data] : detail::zip_equal(names, fields)) {
+    // TODO: Empty name? Double separators?
+    auto segments
+      = name | std::views::split(sep)
+        | std::views::transform([](auto subrange) {
+            return std::string_view{subrange.data(), subrange.size()};
+          });
+    auto current = &root;
+    for (auto segment : segments) {
+      auto record = std::get_if<unflatten_record>(current);
+      if (record) {
+        // nullptr -> all valid. we only want to make more valid.
+        record->null_bitmap
+          = bitmap_or(record->null_bitmap, array.null_bitmap());
+      } else {
+        // TODO: Get null bitmap of all ones?
+        record = &current->emplace<unflatten_record>(array.length(),
+                                                     array.null_bitmap());
+      }
+      current = &record->fields[segment];
+    }
+    // TODO: data might be record
+    unflatten_recurse(*current, data, sep);
+  }
+}
+
+void unflatten_recurse(unflatten_entry& entry,
+                       std::shared_ptr<arrow::Array> array,
+                       std::string_view sep) {
+  if (auto record = caf::get_if<arrow::StructArray>(&*array)) {
+    unflatten_recurse(entry, *record, sep);
+  } else if (auto list = caf::get_if<arrow::ListArray>(&*array)) {
+    entry = unflatten(*list, sep);
+  } else {
+    entry = std::move(array);
+  }
+}
+
+auto realize(unflatten_record&& record) -> std::shared_ptr<arrow::StructArray>;
+
+auto realize(unflatten_entry&& entry) -> std::shared_ptr<arrow::Array> {
+  TENZIR_WARN("pre realize");
+  return entry.match<std::shared_ptr<arrow::Array>>(
+    [](std::shared_ptr<arrow::Array>& array) {
+      TENZIR_ASSERT(array);
+      TENZIR_WARN("realizing {}", array->type()->ToString());
+      return std::move(array);
+    },
+    [](unflatten_record& record) {
+      TENZIR_WARN("realizing record");
```

Comment:
A few leftover debug prints like this, but I'm also reviewing while this is still in draft. Just leaving this one comment here as a reminder.